
# **Import Library**

In [91]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

sns.set(style="whitegrid", font_scale=1.1)
plt.rcParams["figure.dpi"] = 100
pd.set_option("display.max_columns", 20)



---
# **Memuat Dataset**

In [92]:
file = "/content/netflix_titles.csv"
df = pd.read_csv(file)
df.head()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,NaN,United States,"September 25, 2021",2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm..."
1,s2,TV Show,Blood & Water,NaN,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t..."
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",NaN,"September 24, 2021",2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...
3,s4,TV Show,Jailbirds New Orleans,NaN,NaN,NaN,"September 24, 2021",2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo..."
4,s5,TV Show,Kota Factory,NaN,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...


---
# **1. Data Understanding**

**Data Understanding** adalah tahap kedua setelah business understanding dalam metodologi CRISP-DM. Pada tahap ini kita mempelajari struktur, tipe data, dan karakteristik umum dataset sebelum melakukan analisis lebih lanjut.

### Deskripsi Kolom Dataset
| Kolom | Tipe | Deskripsi |
|---|---|---|
| show_id | object | ID unik setiap konten |
| type | object | Jenis konten: Movie atau TV Show |
| title | object | Judul konten |
| director | object | Nama sutradara |
| cast | object | Daftar nama pemain |
| country | object | Negara produksi |
| date_added | object | Tanggal konten ditambahkan ke Netflix |
| release_year | int64 | Tahun rilis konten |
| rating | object | Rating usia penonton (TV-MA, PG-13, dll) |
| duration | object | Durasi (menit untuk film, season untuk series) |
| listed_in | object | Genre konten |
| description | object | Sinopsis singkat |

In [93]:
# Informasi tipe data dan jumlah non-null
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


In [94]:
# Statistik deskriptif semua kolom
df.describe(include='all')

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
count,8807,8807,8807,6173,7982,7976,8797,8807.000000,8803,8804,8807,8807
unique,8807,2,8807,4528,7692,748,1767,NaN,17,220,514,8775
top,s8807,Movie,Zubaan,Rajiv Chilaka,David Attenborough,United States,"January 1, 2020",NaN,TV-MA,1 Season,"Dramas, International Movies","Paranormal activity at a lush, abandoned prope..."
freq,1,6131,1,19,19,2818,109,NaN,3207,1793,362,4
mean,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2014.180198,NaN,NaN,NaN,NaN
std,NaN,NaN,NaN,NaN,NaN,NaN,NaN,8.819312,NaN,NaN,NaN,NaN
min,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1925.000000,NaN,NaN,NaN,NaN
25%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2013.000000,NaN,NaN,NaN,NaN
50%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2017.000000,NaN,NaN,NaN,NaN
75%,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2019.000000,NaN,NaN,NaN,NaN


---
# **2. Verifikasi Kualitas Data**

**Verifikasi kualitas data** adalah proses memastikan data layak dianalisis dengan mengecek:
- Missing values (nilai kosong)
- Inconsistent values (nilai tidak konsisten)
- Duplicate values (data duplikat)
- Outlier values (nilai ekstrem)


## **2.1 Missing Values**

In [95]:
# Persentase missing values per kolom
missing_pct = pd.DataFrame(
    df.isnull().sum() / len(df) * 100,
    columns=['Null Ratio (%)']
).round(2)
missing_pct['Jumlah Null'] = df.isnull().sum()
print(missing_pct)


              Null Ratio (%)  Jumlah Null
show_id                 0.00            0
type                    0.00            0
title                   0.00            0
director               29.91         2634
cast                    9.37          825
country                 9.44          831
date_added              0.11           10
release_year            0.00            0
rating                  0.05            4
duration                0.03            3
listed_in               0.00            0
description             0.00            0


**Interpretasi Missing Values:**
- Kolom `director` memiliki missing value tertinggi sebesar **~29.9%** — hal ini wajar karena banyak konten TV Show yang tidak memiliki satu sutradara spesifik.
- Kolom `cast` dan `country` masing-masing memiliki ~9.4% nilai kosong.
- Kolom penting seperti `title`, `type`, `listed_in`, `release_year`, dan `description` **tidak memiliki missing value (0%)** — data inti aman untuk dianalisis.


## **2.2 Inconsistent Values**

In [96]:
# Cek nilai unik pada kolom rating — ada nilai aneh berupa durasi menit
print("Unique values di kolom 'rating':")
print(df['rating'].unique())
print("\nNilai tidak valid (seharusnya di kolom 'duration'):")
print(df[df['rating'].isin(['74 min', '84 min', '66 min'])][['title', 'type', 'rating', 'duration']])

Unique values di kolom 'rating':
['PG-13' 'TV-MA' 'PG' 'TV-14' 'TV-PG' 'TV-Y' 'TV-Y7' 'R' 'TV-G' 'G'
 'NC-17' '74 min' '84 min' '66 min' 'NR' nan 'TV-Y7-FV' 'UR']

Nilai tidak valid (seharusnya di kolom 'duration'):
                                     title   type  rating duration
5541                       Louis C.K. 2017  Movie  74 min      NaN
5794                 Louis C.K.: Hilarious  Movie  84 min      NaN
5813  Louis C.K.: Live at the Comedy Store  Movie  66 min      NaN


**Temuan:** Terdapat 3 baris di kolom `rating` yang berisi nilai durasi seperti `'74 min'`, `'84 min'`, `'66 min'`. Nilai-nilai ini seharusnya berada di kolom `duration`. Ini adalah contoh **dirty data** yang perlu dibersihkan.


## **2.3 Duplicate Values**

In [97]:
duplikat = df[df.duplicated()]
print(f"Jumlah baris duplikat: {len(duplikat)}")
print("✅ Tidak ada data duplikat!" if len(duplikat) == 0 else duplikat)

Jumlah baris duplikat: 0
✅ Tidak ada data duplikat!


## **2.4 Outlier Values**

In [98]:
# Deteksi outlier pada kolom release_year menggunakan metode IQR
q1 = df['release_year'].quantile(0.25)
q3 = df['release_year'].quantile(0.75)
iqr = q3 - q1
lower = q1 - 1.5 * iqr
upper = q3 + 1.5 * iqr

outliers = df[(df['release_year'] < lower) | (df['release_year'] > upper)]
pct = len(outliers) / len(df) * 100

print(f"Batas bawah IQR : {lower:.0f}")
print(f"Batas atas IQR  : {upper:.0f}")
print(f"Jumlah outlier  : {len(outliers)} ({pct:.2f}%)")
print(f"\nRange release_year: {df['release_year'].min()} - {df['release_year'].max()}")
print("\nCatatan: Outlier berupa film-film lawas (1925-1970an) — valid secara konteks, tidak dihapus.")

Batas bawah IQR : 2004
Batas atas IQR  : 2028
Jumlah outlier  : 719 (8.16%)

Range release_year: 1925 - 2021

Catatan: Outlier berupa film-film lawas (1925-1970an) — valid secara konteks, tidak dihapus.


---
# **3. Preprocessing (Data Cleaning)**

Berdasarkan hasil verifikasi kualitas data, langkah-langkah cleaning yang dilakukan:
1. Perbaiki nilai salah di kolom `rating` (pindahkan nilai durasi ke kolom yang benar)
2. Konversi kolom `date_added` ke format datetime
3. Ekstrak fitur `year_added` dan `month_added`
4. Isi missing values `director` dan `country` dengan `'Unknown'`
5. Ekstrak negara dan genre utama dari kolom multi-nilai


In [99]:


df_clean = df.copy()

# 1. Fix nilai salah di kolom rating
mask_wrong = df_clean['rating'].isin(['74 min', '84 min', '66 min'])
df_clean.loc[mask_wrong, 'duration'] = df_clean.loc[mask_wrong, 'rating']
df_clean.loc[mask_wrong, 'rating'] = np.nan


# 2. Konversi date_added ke datetime
df_clean['date_added'] = pd.to_datetime(df_clean['date_added'].str.strip(), format='%B %d, %Y', errors='coerce')
df_clean['year_added'] = df_clean['date_added'].dt.year
df_clean['month_added'] = df_clean['date_added'].dt.month


# 3. Isi missing values
df_clean['director'] = df_clean['director'].fillna('Unknown')
df_clean['country'] = df_clean['country'].fillna('Unknown')
df_clean['rating'] = df_clean['rating'].fillna('NR')


# 4. Ekstrak primary country dan primary genre
df_clean['primary_country'] = df_clean['country'].str.split(',').str[0].str.strip()
df_clean['primary_genre'] = df_clean['listed_in'].str.split(',').str[0].str.strip()


print(f"\nShape setelah cleaning: {df_clean.shape}")
print(f"Missing values tersisa:")
print(df_clean.isnull().sum()[df_clean.isnull().sum() > 0])


Shape setelah cleaning: (8807, 16)
Missing values tersisa:
cast           825
date_added      10
year_added      10
month_added     10
dtype: int64


# **Menyimpan Dataset**

In [100]:
df.to_csv('netflix_titles(Bersih)', index=False)